crates/ty_python_semantic/src/types/typed_dict.rs

```diff
@@ -246,14 +290,72 @@ impl<'db> TypedDictType<'db> {
         }
         constraints
     }
+
+    pub fn definition(self, db: &'db dyn Db) -> Option<Definition<'db>> {
+        match self {
+            TypedDictType::Class(defining_class) => Some(defining_class.definition(db)),
+            TypedDictType::Synthesized(_) => None,
+        }
+    }
+
+    pub(crate) fn normalized_impl(self, db: &'db dyn Db, visitor: &NormalizedVisitor<'db>) -> Self {
+        match self {
+            TypedDictType::Class(_) => {
+                let synthesized = SynthesizedTypedDictType::new(db, self.items(db));
+                TypedDictType::Synthesized(synthesized.normalized_impl(db, visitor))
+            }
+            TypedDictType::Synthesized(synthesized) => {
+                TypedDictType::Synthesized(synthesized.normalized_impl(db, visitor))
+            }
+        }
+    }
+
+    pub(crate) fn is_equivalent_to_impl(
+        self,
+        db: &'db dyn Db,
+        other: TypedDictType<'db>,
+        inferable: InferableTypeVars<'_, 'db>,
+        visitor: &IsEquivalentVisitor<'db>,
+    ) -> ConstraintSet<'db> {
+        // TODO: `closed` and `extra_items` support will go here. Until then we don't look at the
+        // params at all, because `total` is already incorporated into `FieldKind`.
+
+        // Compare the fields without requiring them to be in sorted order. Class-based `TypedDict`
+        // fields are not sorted. We do sort synthetic fields in `normalized_impl`, but there will
+        // soon be other sources of `SynthesizedTypedDictType` besides normalization.
+        if self.items(db).len() != other.items(db).len() {
+            return ConstraintSet::from(false);
+        }
+        let other_items = other.items(db);
+        self.items(db).iter().when_all(db, |(name, field)| {
+            let Some(other_field) = other_items.get(name) else {
+                return ConstraintSet::from(false);
+            };
+            if field.flags != other_field.flags {
+                return ConstraintSet::from(false);
+            }
+            field
+                .declared_ty
+                .is_equivalent_to_impl(db, other_field.declared_ty, inferable, visitor)
+        })
```

Comment:
I don't think this will change performance much but you could make use of the fact that you have two `BTreeMap` where both should return the keys in the same order if they have the same fields:

```suggestion
        let mut other_items_iter = other_items.iter();

        self_items.iter().when_all(db, |(name, field)| {
            let Some((other_name, other_field)) = other_items_iter.next() else {
                return ConstraintSet::from(false);
            };

            if name != other_name || field.flags != other_field.flags {
                return ConstraintSet::from(false);
            }
            field
                .declared_ty
                .is_equivalent_to_impl(db, other_field.declared_ty, inferable, visitor)
        })
```

Doing the same in `has_relation_to` seems a bit trickier and would require `peek_if`

Reply:
Yes (to the first diff above), and the comment here about not being in sorted order is entirely wrong now :) I think since we're doing a length check up front, we can go ahead and `zip` the iterators, which is a little shorter: 3e5b534c9a64e1a5a09b0ae57b7ca2695849af72